### PR TITLE
Resolve temp playlists within folders

### PR DESCRIPTION
**Pull request by nicoback2**
### Description
Update the temp playlist resolver saga to also resolve temp playlists that are in folders

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?
Unit tests for helpers + tested in browser by creating playlists and moving them around and seeing if they resolved properly

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### CODE REVIEW

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/nkang--resolve-temp-playlists-folders

File: packages/web/src/store/playlist-library/helpers.ts [hunk 1]
```diff
@@ -409,3 +411,53 @@ export const containsTempPlaylist = (
...
+ * with temporary playlists replaced by their resolved playlist identifiers.
+ * @param library
+ * @param tempPlaylistIdToResolvedPlaylist object that maps temporary playlist ids to their resolved playlist identifiers
+ * @returns PlaylistLibrary | PlaylistLibraryFolder
+ */
+export const replaceTempWithResolvedPlaylists = <
+  T extends PlaylistLibrary | PlaylistLibraryFolder
+>(
+  library: T,
+  tempPlaylistIdToResolvedPlaylist: { [key: string]: PlaylistLibraryIdentifier }
```
Comment (sliptype) [inline packages/web/src/store/playlist-library/helpers.ts:447]: TypeScript recommends using the `Record` type now

```suggestion
  tempPlaylistIdToResolvedPlaylist: Record<string, PlaylistLibraryIdentifier>
```

File: packages/web/src/store/playlist-library/sagas.ts [hunk 2]
```diff
@@ -108,23 +110,30 @@ function* watchUpdatePlaylistLibraryWithTempPlaylist() {
...
+    const tempPlaylists = extractTempPlaylistsFromLibrary(playlistLibrary)
+    const resolvedPlaylists: PlaylistLibraryIdentifier[] = yield all(
+      tempPlaylists.map(playlist => call(resolveTempPlaylists, playlist))
     )
-    playlistLibrary.contents = newContents
+    const tempPlaylistIdToResolvedPlaylist = tempPlaylists.reduce(
+      (result, nextTempPlaylist, index) => {
+        result[nextTempPlaylist.playlist_id] = resolvedPlaylists[index]
+        return result
+      },
```
Comment (sliptype) [inline packages/web/src/store/playlist-library/sagas.ts:128]: nit, but you could write this reducer like this to be fully immutable:

```suggestion
      (result, nextTempPlaylist, index) => ({
        ...result,
        [nextTempPlaylist.playlist_id]: resolvedPlaylists[index]
      }),
```

Might be slower though? 🤷

File: packages/web/src/store/playlist-library/helpers.ts [hunk 3]
```diff
@@ -409,3 +411,53 @@ export const containsTempPlaylist = (
...
+ * @param library
+ * @returns PlaylistLibraryIdentifier[]
+ */
+export const extractTempPlaylistsFromLibrary = (
+  library: PlaylistLibrary | PlaylistLibraryFolder
+): PlaylistLibraryIdentifier[] => {
+  if (isEmpty(library.contents)) return []
+  return library.contents.reduce((prevResult, nextContent) => {
+    if (nextContent.type === 'folder') {
+      return prevResult.concat(extractTempPlaylistsFromLibrary(nextContent))
```
Comment (sliptype) [inline packages/web/src/store/playlist-library/helpers.ts:426]: Recursive reduce 😎  sick

**Review by sliptype — APPROVED** [review]
Looks great! Nice work

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/nkang--resolve-temp-playlists-folders

**Pull request merged**